saliency/xrai.py

```diff
@@ -0,0 +1,544 @@
+                 integer_segments=True):
+    """Run approximate XRAI saliency given attributions and segments. This
+       version does not consider mask overlap during importance ranking,
+       significantly speeding up the algorithm for less accurate results.
+
+    Args:
+        attr: Source attributions for XRAI. XRAI attributions will be same size
+              as the input attr.
+        segs: Input segments as a list of boolean masks. XRAI uses these to
+              compute attribution sums.
+        gain_fun: The function that computes XRAI area attribution from source
+                  attributions. Defaults to _gain_density, which calculates the
+                  density of attributions in a mask.
+        area_perc_th: This parameter is ignored. Fast version always computes
+                      to 1.0. It is left here for API compatibility.
+        min_pixel_diff: Do not consider masks that have difference less than
+                        this number compared to the current mask. Set it to 1
+                        to remove masks that completely overlap with the
+                        current mask.
+        integer_segments: See XRAIParameters. Defaults to True.
+
+    Returns:
+        tuple: saliency heatmap and list of masks or an integer image with
+               area ranks depending on the parameter integer_segments.
+    """
+    output_attr = -np.inf * np.ones(shape=attr.shape, dtype=np.float)
+
+    n_masks = len(segs)
+    current_mask = np.zeros(attr.shape, dtype=bool)
+
+    masks_trace = []
+
+    # Sort all masks based on gain, ignore overlaps
+    seg_attrs = [gain_fun(seg_mask, attr) for seg_mask in segs]
+    segs, seg_attrs = list(zip(
+        *sorted(zip(segs, seg_attrs), key=lambda x: -x[1])))
+
+    for i, added_mask in enumerate(segs):
+      mask_diff = _get_diff_mask(added_mask, current_mask)
+      # If mask does not add more than min_pixel_diff to current mask, skip
+      mask_pixel_diff = _get_diff_cnt(added_mask, current_mask)
+      if mask_pixel_diff < min_pixel_diff:
+        if logger.isEnabledFor(logging.DEBUG):
+          logger.debug("Skipping mask with pixel difference: {:.3g},".format(
+              mask_pixel_diff))
+        continue
+      mask_gain = gain_fun(mask_diff, attr)
+      masks_trace.append((mask_diff, mask_gain))
+      output_attr[mask_diff] = mask_gain
+      current_mask = np.logical_or(current_mask, added_mask)
+      if logger.isEnabledFor(logging.DEBUG):
+        current_attr_sum = np.sum(attr[current_mask])
+        current_area_perc = np.mean(current_mask)
+        logger.debug("{} of {} masks processed,"
+                     "attr_sum: {}, area: {:.3g}/{:.3g}".format(
+                         i + 1, n_masks, current_attr_sum, current_area_perc,
+                         area_perc_th))
+    masks_trace = list(zip(*sorted(masks_trace, key=lambda x: -x[1])))[0]
+    uncomputed_mask = output_attr == -np.inf
+    masks_trace.append(uncomputed_mask)
```

Comment:
'masks_trace' is a tuple and calling append() on it will result in crash.

Reply:
ups, changed masks_trace should have been a list. Fixed it!